adjacentSquares.js

```diff
@@ -0,0 +1,105 @@
+/*
+series of squares, 2d plane
+side length of 1
+find size of largest group in adjacent squares
+each sqaure, x-y pair
+
+i: array of squares [[0, 0], [1, 1]]
+o: largest group of adajcent squares
+
+const squares = [
+    [0.5, 0.5],
+    [1.5, 0.5],
+    [2, 0.5],
+    [2.5, -0.5],
+    [2.5, -1.5],
+    [1.5, -1.5],
+];
+
+*/
+
+const SQUARE_SIDE_LENGTH = 1;
+
+const areConected = (nodeOne, nodeTwo) => {
+    const [nodeOneX, nodeOneY] = nodeOne;
+    const [nodeTwoX, nodeTwoY] = nodeTwo;
+
+    const differenceBetweenNodesX = Math.abs(nodeTwoX - nodeOneX);
+    const differenceBetweenNodesY = Math.abs(nodeTwoY - nodeOneY);
+
+    const areEqualedToSideLengthX =
+        differenceBetweenNodesX === 1 && differenceBetweenNodesY < SQUARE_SIDE_LENGTH;
+    const areEqualedToSideLengthY =
+        differenceBetweenNodesY === 1 && differenceBetweenNodesX < SQUARE_SIDE_LENGTH;
+
+    return areEqualedToSideLengthX || areEqualedToSideLengthY;
+};
+
+const buildGraph = (nodes) => {
+    const numNodes = nodes.length;
+    const graph = {};
+
+    for (let nodeOne = 0; nodeOne < numNodes; nodeOne++) {
+        for (let nodeTwo = nodeOne + 1; nodeTwo < numNodes; nodeTwo++) {
+            if (!areConected(nodes[nodeOne], nodes[nodeTwo])) continue;
+
+            const containsNodeOne = graph.hasOwnProperty(nodeOne);
+            const containsNodeTwo = graph.hasOwnProperty(nodeTwo);
+
+            if (!containsNodeOne) graph[nodeOne] = [];
+            if (!containsNodeTwo) graph[nodeTwo] = [];
+
+            graph[nodeOne].push(nodeTwo);
+            graph[nodeTwo].push(nodeOne);
+        }
+    }
+
+    return graph;
+};
+
+const getComponentSize = (startNode, graph, visited) => {
+    const queue = new Queue();
+
+    visited.add(startNode);
+    queue.enqueue(startNode);
+
+    let componentSize = 0;
+    while (queue.size() > 0) {
+        // Remove node
+        const node = queue.dequeue();
+
+        // Process node
+        componentSize++;
+
+        // Get neighbors
+        const containsNode = graph.hasOwnProperty(node);
+        if (!containsNode) continue;
```

Comment:
P: What's the point of this? Why can't we just remove it?

Reply:
The point of this line of code is check if a property exist on the object. Without the check, the code can crash.

Reply:
I need a deeper explanation here. Will `!containsNode` ever evaluate to true? What would cause it to evaluate to true? (i.e. in what scenario would `!containsNode` be true?)

Reply:
Well, since we're working with implicit graphs, there needs to be a relationship between nodes that agree with each other in order to be connected and create our graph (through a Javascript object). But there could be a scenario where the input 2D array doesn't have any nodes that agree with each other, thus our Javascript object can be empty (contains no properties) which will evaluate line 78 to be true.

Reply:
Correct, approved!